abdiff/core/collate_ab_transforms.py

```diff
@@ -167,89 +173,90 @@ def get_transformed_batches_iter(
+    A and B versions of the same intellectual record as a single row, by limiting to the
+    transformed file.
 
-        Steps 2-4 are performed for each transformed file.
+    For performance reasons, DuckDB is provided with ONLY the parquet files that are
+    associated with that transformed file, to prevent it from scanning all files in the
+    dataset (this was memory safe, but somewhat slow given the potentially large number
+    of transformed files).
 
-        2. Using common table expressions (CTEs) execute a query that performs
-           the following:
-           - Create a CTE called 'transformed_file' that contains all the records
-             where the partition transformed_file_name=the name of the transformed file.
-           - Create a CTE called 'a' that contains all the records from CTE
-             'transformed_file' where the column version='a'.
-           - Create a CTE called 'b' that contains all the records from CTE
-             'transformed_file' where the column version='b'.
-           - Join the 'a' and 'b' CTEs using the column timdex_record_id.
+    As records are joined, an "abdiff_record_id" UUID is minted to unambiguously reference
+    that particular row, which is helpful later during deduping.
 
-        3. Read results from the query one batch at a time using fetch_record_batch().
-           Note: This will return the results as a pyarrow.RecordBatchReader.
-
-        4. Yield batches of the joined records until the pyarrow.RecordBatchReader
-           is empty.
-
-    The returned generator can be passed to abdiff.core.utils.write_to_dataset()
-    to perform batch writes of joined TIMDEX A/B records to a Parquet dataset
+    As the records are joined, record batches are yielded which get written to a new,
+    temporary "joined" dataset on disk.
 
     Args:
         dataset_directory: The root directory of the Parquet dataset of TIMDEX records
             (i.e., the tempfile.TemporaryDirectory).
     """
+    partition_paths = glob.glob(f"{dataset_directory}/transformed_file_name=*")
+    transformed_file_names = [os.path.basename(p).split("=")[1] for p in partition_paths]
+
     with duckdb.connect(":memory:") as con:
-        transformed_file_names = con.execute(
-            """
-            SELECT DISTINCT transformed_file_name FROM
-            read_parquet($transformed_parquet_glob, hive_partitioning=true)
-            """,
-            {"transformed_parquet_glob": f"{dataset_directory}/**/*.parquet"},
-        ).fetch_df()["transformed_file_name"]
+        for i, transformed_file in enumerate(transformed_file_names):
+            logger.debug(
+                f"Joining records batch {i+1}/{len(transformed_file_names)}: "
+                f"{transformed_file}"
+            )
+
+            transformed_file_parquet_glob = (
+                f"{dataset_directory}/transformed_file_name={transformed_file}/*.parquet"
+            )
```

Comment:
This is one of the most key updates from this PR.

Instead of providing the full dataset of transformed files to DuckDB and then filtering on a partition of `transformed_file=XYZ`, we limit the actual parquet files DuckDB sees to this partition only.

What's key to point out: **this is NOT more memory efficient**, as DuckDB + parquet are very good at not reading data from parquet files in partitions that the WHERE clause removes, but **it is linearly faster in proportion to how many files DuckDB doesn't waste time just reading the filename**.  

This use of DuckDB + parquet is a bit unusual, as we are looping through potentially thousands of input files and performing a SQL query for each.  It's this loop + that file scan where the delay started to stack up.  If and when we use parquet in TIMDEX, we won't be performing 1-2k SQL queries in a row like this, and so even if we _do_ scan lots of files, it will be a 3-4 second pause before the query returns quickly.

Reply:
Previously, we would pass all parquet files to DuckDB via `.../**/*.parquet` and then rely on DuckDB to filter to the records in quesition via `WHERE transformed_file_name=XYZ`.  

DuckDB was able to keep the other files out of memory via this approach, but there was a time delay as it read the filenames + partitions from the dataset on disk to isolate the parquet files to use (and thus, the records).  So, it was memory-safe, but slower.  

If we were running 1 or 4-5 queries in a row, this wouldn't matter; the delay was small.  But when we're potentially looping through 3.2k+ (1.6k x 2, plus possible more for deletes) file names, this small delay stacks up for each query.

In the new approach, instead of DuckDB filtering the files considered for the query, we do that in advance for the glob pattern of parquet files via `transformed_file_name={transformed_file}/*.parquet`.  Now DuckDB will see like 2-3 files for it's query _and_ no need to filter them based on partition name, so much quicker.